bin/main.rb

```diff
@@ -1,72 +1,203 @@
-  slow_typing(text, 1)
-  loop do
-    answer = gets.chomp!.to_i
-    return answer if (1..9).include? answer
+class Game
+  include QuestionMaker
+  include Visualizer
+  def initialize
+    @board = create_board
+  end
+
+  def player_names
+    @player1 = Player.new(ask('Enter player 1 name:'))
+    temp_name = ask('Enter player 2 name:')
+    while temp_name == @player1.name
+      temp_name = ask("This is the same as player one's name (#{@player1.name}), please enter another name:")
+    end
+    @player2 = Player.new(temp_name)
+    @names = [@player1.name, @player2.name] * 5
+  end
+
+  def play
+    greeting
+    dialog
+    9.times do |turn|
+      show_board(@board)
+      move = question(turn)
+      redo if check_spot(move, turn) == 'jump_next'
+      break if check_game == true
+    end
+    end_message
+  end
+
+  def auto_play(scenario)
+    greeting
+    9.times do |turn|
+      show_board(@board)
+      puts move = scenario[turn]
+      sleep 1
+      redo if check_spot(move, turn) == 'jump_next'
+      break if check_game == true
+    end
+    end_message
+  end
+
+  private
+
+  def dialog
+    dialog_text = @player1.talk_with(@player2)
+    (dialog_text.size / 2).times do |i|
+      print "#{@player1.name}: "
+      print " #{slow_typing(dialog_text[i * 2], 1)}\n"
+      print "#{@player2.name}: "
+      print " #{slow_typing(dialog_text[i * 2 + 1], 1)}\n"
+      slow_typing('...', 20)
+    end
+  end
+
+  def greeting
+    show_board_hint(@board)
+    puts "Welcome to #{@name} game..."
+    player_names
+  end
 
-    puts 'Please enter a number between 1 and 9...'
+  def create_board
+    Array.new(9, 0)
   end
 end
 
-def play
-  greeting
+class TicTacToe < Game
+  def initialize
+    super
+    @name = 'Tic Tac Toe'
+  end
+
+  def to_s
+    'This is a TIC TAC TOE game!'
+  end
+
+  private
+
+  def check_spot(move, turn)
+    if @board[move - 1].zero?
+      make_move(move, turn)
+    else
+      text = "Please make another move. That (#{move}) place is already taken!"
+      slow_typing(text, 0.5)
+      show_board_hint(@board)
+      'jump_next'
+    end
+  end
+
+  def calculate_board
+    sum_of_lines = []
+    sum_of_lines = calculate_linear(sum_of_lines)
+    calculte_diagonal(sum_of_lines)
+  end
+
+  def calculate_linear(array)
+    3.times do |i|
+      sum_line = @board[i * 3] + @board[i * 3 + 1] + @board[i * 3 + 2]
+      array << sum_line
+      sum_line = @board[i] + @board[i + 3] + @board[i + 6]
+      array << sum_line
+    end
+    array
+  end
 
-  board = create_board
+  def calculte_diagonal(array)
+    2.times do |i|
+      sum_diagonal = @board[i + i] + @board[4] + @board[8 - i * 2]
+      array << sum_diagonal
+    end
+    array
+  end
+
+  def make_move(move, turn)
+    mark = turn.even? ? 1 : 10
+    @board[move - 1] = mark
+  end
+
+  def check_game
+    return true if calculate_board.include?(30)
+    return true if calculate_board.include?(3)
+    return true unless @board.include?(0)
+  end
```

Comment:
- **[SUGGESTION]** I'd recommend implementing the check_game logic as follows, it's clean, understandable and maintainable.

```ruby
class Game
  board = ['O', 'X', 'O', 'X', 'O', 'X', 'O', 'X', 'X']

  # bad
  def check_win
    # for rows
    return 'X' if board[0] == board[1] == board[2] == 'X'
    return 'O' if board[0] == board[1] == board[2] == 'O'
    return 'X' if board[3] == board[4] == board[5] == 'X'
    return 'O' if board[3] == board[4] == board[5] == 'O'
    return 'X' if board[6] == board[7] == board[8] == 'X'
    return 'O' if board[6] == board[7] == board[8] == 'O'

    # and analogical as above for columns and diagonals
    # which makes it really long...

  end

  # better
  def check_win(player_symbol)
    # for rows
    all_equal(board[0..2], player_symbol) ||
    all_equal(board[3..5], player_symbol) ||
    all_equal(board[6..8], player_symbol)

    # and analogical as above for columns and diagonals
  end

  def all_equal(subarray, player_symbol)
    return true if subarray.uniq.size == 1 && subarray.uniq == player_symbol
  end
end
```